runtime/sema/check_expression.go

```diff
@@ -41,16 +41,39 @@ func (checker *Checker) VisitIdentifierExpression(expression *ast.IdentifierExpr
 
 	checker.checkSelfVariableUseInInitializer(variable, identifier.Pos)
 
+	checker.checkReferenceValidity(variable, expression)
+
 	if checker.inInvocation {
 		checker.Elaboration.IdentifierInInvocationTypes[expression] = valueType
 	}
 
 	return valueType
 }
 
+func (checker *Checker) checkReferenceValidity(variable *Variable, hasPosition ast.HasPosition) {
+	typ := UnwrapOptionalType(variable.Type)
+	if _, ok := typ.(*ReferenceType); !ok {
+		return
+	}
+
+	referencedVar := checker.references.Find(variable.Identifier)
+	if referencedVar == nil ||
+		!referencedVar.Type.IsResourceType() {
+		return
+	}
+
+	resourceInfo := checker.resources.Get(Resource{Variable: referencedVar})
+	if resourceInfo.Invalidations.Size() == 0 {
+		return
+	}
+
+	checker.report(&InvalidatedResourceReferenceError{
+		Range: ast.NewRangeFromPositioned(checker.memoryGauge, hasPosition),
```

Comment:
Maybe also include the position of the invalidation to aid the developer.

See [the implementation of `ResourceUseAfterInvalidationError` on `master`](https://github.com/onflow/cadence/blob/d8b6241606b9dd78a5ca5d9b646dfa27f7842037/runtime/sema/errors.go#L1869-L1909).

Reply:
Seems I need to wait for this PR https://github.com/onflow/cadence/pull/2051

Reply:
You don't have to wait for it necessarily, what I referred to was just meant as inspiration.

Reply:
I could re-use the same `ErrorNote` interface. But the API has changed a bit in the master, so the two PRs might conflict. I can wait for that to merge, or also can add this in a follow-up PR.

Reply:
Ah, I see. That's a good plan!